src/Build/Definition/ProjectCollection.cs

```diff
@@ -719,34 +706,18 @@ public HostServices HostServices
         {
             get
             {
-                // Avoid write lock if possible, this getter is called a lot during Project construction.
-                using (_locker.EnterDisposableUpgradeableReadLock())
+                HostServices hostServices = Volatile.Read(ref _hostServices);
+                if (hostServices != null)
                 {
-                    if (_hostServices != null)
-                    {
-                        return _hostServices;
-                    }
-
-                    using (_locker.EnterDisposableWriteLock())
-                    {
-                        return _hostServices ?? (_hostServices = new HostServices());
-                    }
+                    return hostServices;
                 }
+                hostServices = new HostServices();
+                return Interlocked.CompareExchange(ref _hostServices, hostServices, null) ?? hostServices;
             }
 
             set
             {
-                bool sendEvent = false;
-                using (_locker.EnterDisposableWriteLock())
-                {
-                    if (_hostServices != value)
-                    {
-                        _hostServices = value;
-                        sendEvent = true;
-                    }
-                }
-
-                if (sendEvent)
+                if (Interlocked.Exchange(ref _hostServices, value) != value)
```

Comment:
Calling `Interlocked.Exchange` to avoid multiple `OnProjectCollectionChanged` callbacks if this is run concurrently.